tket2-hseries/src/extension/wasm.rs

```diff
@@ -0,0 +1,885 @@
+        };
+        CustomType::new(
+            FUNC_TYPE_NAME.to_owned(),
+            [row_to_arg(inputs.into()), row_to_arg(outputs.into())],
+            EXTENSION_ID,
+            TypeBound::Copyable,
+            extension_ref,
+        )
+    }
+
+    fn custom_type(&self, extension_ref: &Weak<Extension>) -> CustomType {
+        match self {
+            Self::Module => CustomType::new(
+                MODULE_TYPE_NAME.to_owned(),
+                [],
+                EXTENSION_ID,
+                TypeBound::Copyable,
+                extension_ref,
+            ),
+            Self::Context => CustomType::new(
+                CONTEXT_TYPE_NAME.to_owned(),
+                [],
+                EXTENSION_ID,
+                TypeBound::Any,
+                extension_ref,
+            ),
+            s @ Self::Func { .. } => s.clone().into_custom_type(extension_ref),
+        }
+    }
+
+    fn into_custom_type(self, extension_ref: &Weak<Extension>) -> CustomType {
+        match self {
+            Self::Module | Self::Context => self.custom_type(extension_ref),
+            Self::Func { inputs, outputs } => {
+                Self::func_custom_type(inputs, outputs, extension_ref)
+            }
+        }
+    }
+
+    fn into_type(self, extension_ref: &Weak<Extension>) -> Type {
+        self.into_custom_type(extension_ref).into()
+    }
+}
+
+impl From<WasmType> for CustomType {
+    fn from(value: WasmType) -> Self {
+        value.into_custom_type(&EXTENSION_REF)
+    }
+}
+
+impl From<WasmType> for Type {
+    fn from(value: WasmType) -> Self {
+        value.into_type(&EXTENSION_REF)
+    }
+}
+
+impl TryFrom<Type> for WasmType {
+    type Error = ();
+
+    fn try_from(value: Type) -> Result<Self, Self::Error> {
+        let TypeEnum::Extension(custom_type) = value.as_type_enum() else {
+            Err(())?
+        };
+
+        custom_type.to_owned().try_into()
+    }
+}
+
+impl TryFrom<CustomType> for WasmType {
+    type Error = ();
+    fn try_from(value: CustomType) -> Result<Self, Self::Error> {
+        if value.extension() != &EXTENSION_ID {
+            Err(())?
+        }
+
+        match value.name() {
+            n if *n == *MODULE_TYPE_NAME => Ok(WasmType::Module),
+            n if *n == *CONTEXT_TYPE_NAME => Ok(WasmType::Context),
+            n if *n == *FUNC_TYPE_NAME => {
+                let [inputs, outputs] = value.args() else {
+                    Err(())?
+                };
+                let inputs = type_arg_into_type_row_rv(inputs.clone()).ok_or(())?;
+                let outputs = type_arg_into_type_row_rv(outputs.clone()).ok_or(())?;
+
+                Ok(WasmType::Func { inputs, outputs })
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+/// Fallibly convert a [TypeArg] to a [Type].
+///
+/// This will fail if `arg` is of non-type kind (e.g. String),
+/// or if `arg` contains row variables.
+/// TODO move this to `impl TypeArg` in `hugr-core`
```

Comment:
Are you leaving this here for the moment because you do not want to make changes to `hugr-core` right now?